sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -260,24 +260,42 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       import org.apache.spark.sql.execution.datasources.v2.DataSourceV2Implicits._
       provider.getTable(dsOptions) match {
         case table: SupportsWrite if table.supports(BATCH_WRITE) =>
-          if (partitioningColumns.nonEmpty) {
-            throw new AnalysisException("Cannot write data to TableProvider implementation " +
-              "if partition columns are specified.")
-          }
           lazy val relation = DataSourceV2Relation.create(table, dsOptions)
           mode match {
             case SaveMode.Append =>
+              checkPartitioningMatchesV2Table(table)
```

Comment:
Append and overwrite should also use `SupportsCatalogOptions`, right? I think we want to always use the catalog to load a table if `SupportsCatalogOptions` is available.

Reply:
We can probably do this in a follow-up. What do you think, @brkyvz?